### PR TITLE
fix: new sorting algorithm

### PR DESCRIPTION
**Pull request by MaxGraey**
New sorting algorithm based on "Power Sort" from "Nearly-Optimal Mergesorts" (May 2018) paper (https://arxiv.org/pdf/1805.04154.pdf). Also this implementation contains highly optimized novel insertion sort as part of main algorithm.

The previous algorithm was [unstable](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) which forced us to do fallback on the slow insertion sort for reference types for following ECMAScript spec. Which, of course, was very undesirable.

Features:

- Stable
- Faster than TimSort (approx 1.5-2x)
- Faster than JS sort (TimSort) [(approx 5x)](https://github.com/AssemblyScript/assemblyscript/pull/1904#issuecomment-869538612)
- Has a mathematical proof of correctness, unlike TimSort (See paper)

- [x] I've read the contributing guidelines

### CODE REVIEW

**Comment by MaxGraey** [comment]
Benchmark: https://webassembly.studio/?f=8ehyqora2yn

Results (Chrome 91):

AS sort 100_000 doubles : 20.30 ms 
AS sort 10_000_000 doubles: 5550.30 ms 

**AS NEW sort 100_000 doubles : 11.60 ms**
**AS NEW sort 10_000_000 doubles: 1718.90 ms**

JS sort 100_000 doubles : 44.90 ms 
JS sort 10_000_000 doubles: 8140.10 ms

**Comment by JairusSW** [comment]
Yayy!

File: std/assembly/util/sort.ts [hunk 1]
```diff
@@ -1,5 +1,18 @@
 import { compareImpl } from "./string";
 
+// @ts-ignore: decorator
+@lazy const EMPTY: u32 = 0xFFFFFFFF;
+// @ts-ignore: decorator
+@lazy const INSERTION_SORT_THRESHOLD = 128;
+// @ts-ignore: decorator
+@lazy const MIN_RUN_LENGTH = 32;
```
Comment (dcodeIO) [inline std/assembly/util/sort.ts:8]:
```suggestion
// @ts-ignore: decorator
@inline const EMPTY: u32 = 0xFFFFFFFF;
// @ts-ignore: decorator
@inline const INSERTION_SORT_THRESHOLD = 128;
// @ts-ignore: decorator
@inline const MIN_RUN_LENGTH = 32;
```

File: std/assembly/util/sort.ts [hunk 2]
```diff
@@ -43,100 +56,241 @@ export function COMPARATOR<T>(): (a: T, b: T) => i32 {
...
+        case 3: {
+          let a = load<T>(dataStart, 1 << alignof<T>());
+          let b = load<T>(dataStart, 2 << alignof<T>());
+          let c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 1 << alignof<T>());
+          b = select<T>(b, a, c >>> 31);
+          a = load<T>(dataStart, 0);
+          c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 0);
+          store<T>(dataStart, select<T>(b, a, c >>> 31), 2 << alignof<T>());
```
Comment (dcodeIO) [inline std/assembly/util/sort.ts:80]:
```suggestion
          let c = comparator(a, b) < 0;
          store<T>(dataStart, select<T>(a, b, c), 1 << alignof<T>());
          b = select<T>(b, a, c);
          a = load<T>(dataStart, 0);
          c = comparator(a, b) < 0;
          store<T>(dataStart, select<T>(a, b, c), 0);
          store<T>(dataStart, select<T>(b, a, c), 2 << alignof<T>());
```

File: std/assembly/util/sort.ts [hunk 3]
```diff
@@ -43,100 +56,241 @@ export function COMPARATOR<T>(): (a: T, b: T) => i32 {
...
+        }
+        case 2: {
+          let a = load<T>(dataStart, 0);
+          let b = load<T>(dataStart, 1 << alignof<T>());
+          let c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 0);
+          store<T>(dataStart, select<T>(b, a, c >>> 31), 1 << alignof<T>());
```
Comment (dcodeIO) [inline std/assembly/util/sort.ts:87]:
```suggestion
          let c = comparator(a, b) < 0;
          store<T>(dataStart, select<T>(a, b, c), 0);
          store<T>(dataStart, select<T>(b, a, c), 1 << alignof<T>());
```

File: std/assembly/util/sort.ts [hunk 4]
```diff
@@ -43,100 +56,241 @@ export function COMPARATOR<T>(): (a: T, b: T) => i32 {
...
-      let b: T = load<T>(dataStart + (<usize>j << alignof<T>())); // b = arr[j]
-      if (comparator(a, b) < 0) {
-        store<T>(dataStart + (<usize>(j-- + 1) << alignof<T>()), b); // arr[j + 1] = b
-      } else break;
-    }
-    store<T>(dataStart + (<usize>(j + 1) << alignof<T>()), a); // arr[j + 1] = a
+  var lgPlus2         = log2u(n) + 2;
+  var lgPlus2Size     = lgPlus2 << alignof<u32>();
+  var leftRunStartBuf = __alloc(lgPlus2Size);
+  var leftRunEndBuf   = __alloc(lgPlus2Size);
```
Comment (dcodeIO) [inline std/assembly/util/sort.ts:100]: Can this share a single allocation of double the size? Something like:

```suggestion
  var leftRunStartBuf = __alloc(2 * lgPlus2Size);
  var leftRunEndBuf   = leftRunStartBuf + lgPlus2Size;
```

**Review by dcodeIO — COMMENTED** [review]
Can't really comment on the algorithm yet, but have a few more general comments :)

File: std/assembly/util/sort.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/util/sort.ts:80]: Originally it was `x < 0` but I figured out `x >>> 31` approx 5-7% faster

Comment (dcodeIO) [reply std/assembly/util/sort.ts:80]: Hmm, 5-7% seems too much for something like that. What happens with `comparator(a, b) >>> 31`?

Comment (MaxGraey) [reply std/assembly/util/sort.ts:80]: I guess we could add `TODO` about this and refactor to `x < 0` in the future when binaryen will support this peephole optimization. Unfortunately my previous attempts [was fail](https://github.com/WebAssembly/binaryen/pull/2868)

Comment (dcodeIO) [reply std/assembly/util/sort.ts:81]: So, what if just the assignment of `c` is changed to `comparator(a, b) >>> 31`? That'd replace four shifts with just one at least.

**Review by dcodeIO — APPROVED** [review]

**Pull request merged**